text/3723-static-function-argument-unpacking.md

```diff
@@ -0,0 +1,882 @@
+
+Since this is dead code, a lint is emitted.
+
+### Generic Parameters
+
+When function parameters are generic, using `<T>`, `impl` or `dyn`, exactly the same should happen as when the arguments are passed by hand. I.e., the argument's type must be *compatible* with the parameter's type. Just as when entering that argument manually with a field access expression.
+
+### References and Mutability in Function Parameters
+
+Rust's current syntax for calling functions that take one parameter by value, by reference, or by mutable reference requires the developer to be explicit about what they want:
+```rust
+fn ret_one_arg() -> u8 {
+    123
+}
+
+fn use_one_arg(p: u8) {
+    println!("{p}");
+}
+
+fn use_one_refarg(p: &u8) {
+    println!("{p}");
+}
+
+fn use_one_refmutarg(p: &mut u8) {
+    println!("{p}");
+}
+
+fn main() {
+    use_one_arg(ret_one_arg());
+    use_one_refarg(&ret_one_arg()); // `&` needed to compile!
+    use_one_refmutarg(&mut ret_one_arg()); // `&mut` needed to compile!
+}
+```
+
+Rust also requires the developer to explicitly dereference references:
+```rust
+const CONST_NUMBER: u8 = 42;
+
+fn ret_one_refarg() -> &'static u8 {
+    &CONST_NUMBER
+}
+
+fn use_one_arg(p: u8) {
+    println!("{p}");
+}
+
+fn use_one_refarg(p: &u8) {
+    println!("{p}");
+}
+
+fn main() {
+    use_one_arg(*ret_one_refarg()); // `*` needed to compile!
+    use_one_refarg(ret_one_refarg());
+}
+```
+
+Explicitly indicating varying degrees of (de)reference status or mutability on arguments being unpacked does not follow from the proposed syntax in any straightforward way. Thus, although it limits the usefulness of the feature, the design for such possibility is left out of the scope of this proposal. Consequently, the code will only compile if passing the arguments one by one with the corresponding field access expressions would compile.
+
+### Type Coercions and Conversions of Collections
+
+If the expression being unpacked is a reference to one of the allowed collection types (tuple, tuple struct, or fixed-size array), whether argument unpacking compiles depends on if accessing **all** its fields manually with the field access expression (`.idx`, or `[idx]`) would work at compile time. If it does, then argument unpacking works. (For the reference, see [`std::ops::Deref`](https://doc.rust-lang.org/std/ops/trait.Deref.html) and [type coercions](https://doc.rust-lang.org/reference/type-coercions.html).) Attempts to access nonexistent fields of tuples and fixed-size arrays or references thereof, using indices past the types' ranges, lead to compilation errors. With argument unpacking, only the fields known of during program compilation are unpacked.
+
+For example, the following compiles, since the alternative compiles:
+```rust
+fn consume(a: u8, b: u8, c: u8) {
+    println!("{a}, {b}, {c}");
+}
+
+fn main() {
+    let tup = &(1, 2, 3);
+    consume(...tup);
+    // Alternative: consume(tup.0, tup.1, tup.2);
+}
+```
```

Comment:
What does `...foo` do if `foo` is a tuple struct that also `Deref`s to a different tuple struct? And does the answer change based on whether `foo`’s fields are visible at the location?

More broadly, wile this autoderef behavior *may* be a good idea, it needs justification. Why should this be in the MVP? Why is `consume(...*tup);` inadequate?